tools/PI/DevHome.PI/BarWindowVertical.xaml.cs

```diff
@@ -0,0 +1,318 @@
+    private HWINEVENTHOOK focusEventHook;
+
+    internal HWND ThisHwnd { get; private set; }
+
+    public Microsoft.UI.Dispatching.DispatcherQueue TheDispatcher
+    {
+        get; set;
+    }
+
+    public BarWindowVertical(BarWindowViewModel model)
+    {
+        viewModel = model;
+
+        // The main constructor is used in all cases, including when there's no target window.
+        TheDispatcher = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();
+
+        InitializeComponent();
+        viewModel.PropertyChanged += ViewModel_PropertyChanged;
+        winPositionEventDelegate = new(WinPositionEventProc);
+        winFocusEventDelegate = new(WinFocusEventProc);
+    }
+
+    private void ViewModel_PropertyChanged(object? sender, PropertyChangedEventArgs e)
+    {
+        if (e.PropertyName == nameof(BarWindowViewModel.IsSnapped))
+        {
+            if (viewModel.IsSnapped)
+            {
+                Snap();
+            }
+            else
+            {
+                Unsnap();
+            }
+        }
+    }
+
+    private void MainPanel_Loaded(object sender, RoutedEventArgs e)
+    {
+        ThisHwnd = (HWND)WindowNative.GetWindowHandle(this);
+
+        // Apply the user's chosen theme setting.
+        ThemeName t = ThemeName.Themes.First(t => t.Name == settings.CurrentTheme);
+        SetRequestedTheme(t.Theme);
+
+        // Regardless of what is set in the XAML, our initial window width is too big. Setting this to 70 (same as the XAML file)
+        Width = 70;
+    }
+
+    private void ExternalToolButton_Click(object sender, RoutedEventArgs e)
+    {
+        if (sender is Button clickedButton)
+        {
+            if (clickedButton.Tag is ExternalTool tool)
+            {
+                var process = tool.Invoke(TargetAppData.Instance.TargetProcess?.Id, TargetAppData.Instance.HWnd);
+
+                if (process == null)
+                {
+                    // It appears ContentDialogs only render in the space it's parent occupies. Since the parent is a narrow
+                    // bar, the dialog doesn't have enough space to render. So, we'll use MessageBox to display errors.
+                    PInvoke.MessageBox(
+                        ThisHwnd,
+                        string.Format(CultureInfo.CurrentCulture, errorMessageText, tool.Executable),
+                        errorTitleText,
+                        MESSAGEBOX_STYLE.MB_ICONERROR);
+                }
+            }
+        }
+    }
+
+    private void WindowEx_Closed(object sender, WindowEventArgs args)
+    {
+        TargetAppData.Instance.ClearAppData();
+
+        if (positionEventHook != IntPtr.Zero)
+        {
+            PInvoke.UnhookWinEvent(positionEventHook);
+            positionEventHook = HWINEVENTHOOK.Null;
+        }
+
+        if (focusEventHook != HWINEVENTHOOK.Null)
+        {
+            PInvoke.UnhookWinEvent(focusEventHook);
+            focusEventHook = HWINEVENTHOOK.Null;
+        }
```

Comment:
Should call 
var primaryWindow = Application.Current.GetService<PrimaryWindow>();
        primaryWindow.ClearBarWindow();

Can we extend BarWindowVertical and BarWindowHorizontal from BarWindow, so we can have a common WindowEx_Closed or call a common method in viewmodel?

Reply:
Good catch. Yes, added a comment method.